csharp/rust/src/buffering.rs

```diff
@@ -0,0 +1,125 @@
+impl FFIBuffer {
+    pub fn new() -> Self {
+        Self {
+            small_buffer: [0; SMALL_BUFFER_SIZE],
+            small_buffer_index: 0,
+            large_buffer: Vec::new(),
+            buffer_size_need: 0,
+            count_mode: true,
+            is_large_buffer_active: false,
+        }
+    }
+
+    /// Changes the mode into count_mode, allowing buffer usage.
+    pub fn switch_mode(&mut self) {
+        assert!(self.count_mode);
+        self.count_mode = false;
+        if self.buffer_size_need > SMALL_BUFFER_SIZE {
+            self.is_large_buffer_active = true;
+            self.large_buffer = Vec::with_capacity(self.buffer_size_need);
+        } else {
+            self.is_large_buffer_active = false;
+        }
+        self.buffer_size_need = 0;
+        self.small_buffer_index = 0;
+    }
+
+    /// # Summary
+    /// Writes to the buffer, returning a pointer into the buffer if count_mode is false
+    ///
+    /// # Note
+    /// This struct has state! See the struct documentation for usage.
+    ///
+    /// # Regarding safety
+    /// This method is not safe for use if used incorrectly!
+    /// The stateflow MUST always be kept!
+    pub fn write_to_buffer(&mut self, data: &[u8]) -> *const u8 {
+        if self.count_mode {
+            self.buffer_size_need += data.len();
+            null()
+        } else {
+            unsafe {
+                if self.is_large_buffer_active {
+                    let ptr = self.large_buffer.as_ptr().add(self.large_buffer.len());
+                    self.large_buffer.extend_from_slice(data);
+                    ptr
+                } else {
+                    let ptr = self.small_buffer.as_mut_ptr().add(self.small_buffer_index);
+                    self.small_buffer[self.small_buffer_index..data.len()].copy_from_slice(data);
+                    self.small_buffer_index += data.len();
+                    ptr
+                }
+            }
+        }
+    }
+
+    pub fn write_values_to_buffer(&mut self, data: &[Value]) -> *const u8 {
+        let data = unsafe { helpers::any_as_u8_slice(&data) };
+        self.write_to_buffer(data)
+    }
+    pub fn write_string_pair_to_buffer(&mut self, first: &str, second: &str) -> *const u8 {
+        unsafe {
+            let a = self.write_to_buffer(first.as_bytes()) as *mut c_void;
+            let b = self.write_to_buffer(second.as_bytes()) as *mut c_void;
+            debug_assert!((a.is_null() && b.is_null()) || a.add(first.len()) == b);
+            let output = StringPair {
+                a_start: a as *mut c_char,
+                a_end: a.add(first.len()) as *mut c_char,
+                b_start: b as *mut c_char,
+                b_end: b.add(second.len()) as *mut c_char,
+            };
+            let data = helpers::any_as_u8_slice(&output);
+            self.write_to_buffer(data)
+        }
+    }
+
+    pub fn write_string_to_buffer(&mut self, data: &CString) -> *const u8 {
```

Comment:
We're working mostly with strings which are `byte[]`, so they represented by 2 params: `*const u8` and `u32` (len).